### PR TITLE
Fixes shard level snapshot metadata loading when index-N file is missing

### PR DESCRIPTION
**Pull request by abeyad**
In making changes for the 5.0 version of snapshots, a bug was
introduced where if an index-N file could not be found for an
individual shard, the backup was to iterate over all `snap-*.dat`
files in the shard folder to know which snapshots contain that
shard's data, but in 5.0, reading the `snap-*.dat` files as backup
was incorrectly passing in the blob name for the `snap-*.dat` file,
thereby failing to load all index files for a given snapshot
when the index-N file is missing.  This condition should be rare
as there is no reason an index-N file should be absent (unless
it was deleted or there was corruption reading the file), but
nevertheless, this situation can be encountered and this commit
fixes the bug by reading the correct `snap-*.dat` blob name in the
shard data folder.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -1158,15 +1159,18 @@ protected long findLatestFileNameGeneration(Map<String, BlobMetaData> blobs) {
                     final String file = SNAPSHOT_INDEX_PREFIX + latest;
                     logger.warn((Supplier<?>) () -> new ParameterizedMessage("failed to read index file [{}]", file), e);
                 }
+            } else if (blobKeys.isEmpty() == false) {
+                logger.debug("Could not find a readable index-N file in a non-empty shard snapshot directory [{}]",
+                    () -> blobContainer.path());
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: why lambda here?

**Review by ywelsch — APPROVED** [review]
Left one minor comment (I think it was an oversight). Once addressed, LGTM.

File: core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 1 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: you're right no lamda needed here, as the `path` is already computed and stored as an instance variable on `AbstractBlobContainer` which all implementations inherit from, so no advantage to a lamda here.  I will change it.

**Comment by abeyad** [comment]
5.x commit: 8ae87b2d02e13bbc266baa96d26192c9c81fb426
5.1 commit: 0b4dd4a23ebf3c80f7dfe463861676f7b106f4f2
5.0 commit: 01640a676e6d7b5a4b9670304f436d4f1dcded7b

**Comment by abeyad** [comment]
thanks for the review @ywelsch

**Pull request merged**